### PR TITLE
Issue #5250 

### PR DESCRIPTION
**Pull request by bck5**
Add a new Provider for cardinal directions to solve issue #5250 (Compass not working on Moto G)

### CODE REVIEW

File: main/src/cgeo/geocaching/sensors/MagnetometerAndAccelerometerProvider.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+            private boolean lastMagnetometerSet = false;
+            private final float[] rotateMatrix = new float[9];
+            private final float[] orientation = new float[3];
+            @Override
+            public void call(final Subscriber<? super Float> subscriber) {
+                final SensorEventListener listener = new SensorEventListener() {
+                    @Override
+                    public void onSensorChanged(final SensorEvent sensorEvent) {
+
+                        if(sensorEvent.sensor.equals(accelerometerSensor)) {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/sensors/MagnetometerAndAccelerometerProvider.java]: Please use space after `if`.

File: main/src/cgeo/geocaching/sensors/MagnetometerAndAccelerometerProvider.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    private MagnetometerAndAccelerometerProvider() {
+        // Utility class, not to be instantiated
+    }
+
+    public static Observable<Float> create(final Context context) {
+
+        final SensorManager sensorManager = (SensorManager) context.getSystemService(Context.SENSOR_SERVICE);
+        final Sensor accelerometerSensor = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
+        final Sensor magenetometerSensor = sensorManager.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD);
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/sensors/MagnetometerAndAccelerometerProvider.java]: Typo (should be `magnetometerSensor`).

File: main/src/cgeo/geocaching/sensors/Sensors.java [hunk 3]
```diff
@@ -100,7 +102,16 @@ public void setupDirectionObservable() {
 
         // On some devices, the orientation sensor (Xperia and S4 running Lollipop) seems to have been deprecated for real.
         // Use the rotation sensor if it is available unless the orientatation sensor is forced by the user.
-        final Observable<Float> sensorDirectionObservable = Settings.useOrientationSensor(app) ? OrientationProvider.create(app) : RotationProvider.create(app);
+        // after updating Moto G there is no rotation sensor anymore. Use magnetic field and accelerometer instead.
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/sensors/Sensors.java]: Please use upper-case `a` at the beginning of the sentence.

**Comment by samueltardieu** [comment]
The patch looks mostly fine to me. There should be some information in `SystemInformation.java` too.

**Comment by samueltardieu** [comment]
Also, please use a terser first line in the commit message, such as:

> fix #5250: compass not working on Moto G
>
> Add a new Provider for cardinal directions.

**Comment by bck5** [comment]
I corrected your remarks and also add information about the used sensors in `SystemInformation.java` .

I also used another (better?) commit message, hope it is ok this time.

**Comment by Lineflyer** [comment]
@bck5 
I am taking care of testing and support for c:geo (as far as I can).
Thus I would be glad if you could provide some explanation about the Moto G problem and how you solved it in "some easy words" to have a basis for testing and feedback to users.
What is the difference in the sensor meachnism and to what devices does it apply?
Is this working around a problem on the devices or a general new approach?

**Comment by bck5** [comment]
The Problem is: after the update for Moto G it seems there is no more rotation sensor provided by the system.It also seems the system do not offer the orientation sensor. Cgeo use one of these sensors, to get the direction right.

But these sensors are not the only one for calculating the cardinal directions. With help of the magnetic field sensor and the accelerometer it is also possible to get the right rotation matrix.[1]

So my fix is a new approach calculating the rotation matrix. The sensor type "TYPE_ROTATION_VECTOR" which is used by cgeo can be hardware or software based (see [2]). Because of this I think it was only a software sensor which does not exists anymore after the Moto G update.  

This fix will give all device the possibility to use the compass if they provided the sensor types "TYPE_MAGNETIC_FIELD" and "TYPE_ACCELEROMETER". These both sensors are supported by android since version 1.5[2]. I don't know which devices provided these two, but not rotation_vector or orientation sensor. If there exists such devices (like the Moto G) this fix will apply to them, too.


To say it in easier words: my fix (more or less) emulates the rotation vector sensor and doing this it brings the compass back. 

hope this helps
bck5


[1]https://developer.android.com/reference/android/hardware/SensorManager.html#getRotationMatrix(float[], float[], float[], float[])
[2]https://developer.android.com/guide/topics/sensors/sensors_overview.html

**Comment by samueltardieu** [comment]
Thanks for your contribution. I will first merge it to `master` and let people field test it for a few days before we merge it to release.

**Comment by samueltardieu** [comment]
Closing the pull request (merged to `master`). Thanks for basing your work on `release` as expected for a bug fix, this will let us put it in this branch after the field test.

**Comment by Lineflyer** [comment]
@bck5 

> This fix will give all device the possibility to use the compass if they provided the sensor types "TYPE_MAGNETIC_FIELD" and "TYPE_ACCELEROMETER". These both sensors are supported by android since version 1.5[2]. I don't know which devices provided these two, but not rotation_vector or orientation sensor. If there exists such devices (like the Moto G) this fix will apply to them, too.

Thanks for explaining...
What is the priority in which the different approaches are now used... or more practical:
Will this change also affect devices which do have the "normal" sensors?

**Comment by bck5** [comment]
It will not.

If devices have the "old" sensors, it will use them, or if the orientation sensor is forced it will use the orientation sensor.
So it will behave like it does before the patch.

**Comment by samueltardieu** [comment]
The orientation sensor is deprecated but may result in a better fusion (for example, it may use the gyroscope if present). Is it absent or is it non-working on Moto G?

**Comment by samueltardieu** [comment]
Ok, from the info I got from a friend trying the NB, it is present but non-working.

```
--- System information ---
Device: MotoG3 (cm_osprey, Motorola)
Android version: 5.1.1
Android build: cm_osprey-userdebug 5.1.1 LMY48W 9030813075 test-keys
Cgeo version: 2015.11.08-NB-4482e03
Google Play services: enabled
Low power mode: active
Compass capabilities: yes
Rotation vector sensor: absent
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: magnetometer & accelerometer
Hide own/found: false
Map strategy: auto
HW acceleration: enabled (default state)
System language: fr_FR
Geocaching sites enabled:
- geocaching.com: logged in (Connecté) / Basic
Installed cgeo plugins:  none
--- End of system information ---

Maybe we should give the user a choice of the sensor to use in the settings rather than a simple "force use of rotation sensor".
```

**Comment by bck5** [comment]
Yes, I run a stock Android on the MotoG1.
```
--- System information ---
Device: XT1032 (falcon_retde, motorola)
Android version: 5.1
Android build: LPB23.13-56
Cgeo version: 2015.11.10-NB-477e26c
Google Play services: enabled
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: absent
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: magnetometer & accelerometer
Hide own/found: false
Map strategy: auto
HW acceleration: enabled (default state)
System language: de_DE
Geocaching sites enabled:
- geocaching.com: not logged in (Keine Anmeldedaten gespeichert.)
Installed cgeo plugins:  none
--- End of system information ---
```

The system info says, the orientation sensor is present, but using the orientation sensor for the compass results in a non-working compass.

**Comment by samueltardieu** [comment]
By non-working, you mean that it outputs no value at all?

**Comment by bck5** [comment]
I did not looked at the values it outputs, but it seems it is always outputs the same value. It results in a non moving compass.

**Pull request closed without merging**